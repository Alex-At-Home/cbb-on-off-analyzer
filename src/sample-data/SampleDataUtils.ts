import _ from "lodash";

export class SampleDataUtils {

  /** For snapshot testing, provides a more useful output than a giant list of stats */
  static summarizeEnrichedApiResponse(exampleRawResponse: Record<string, any>, mutableSample: boolean = false) {
    const expectedFields1 = mutableSample ? null : new Set(_.keys(exampleRawResponse));

    return {
      test: (val: any) => val && (val.off_efg ||
          val.hasOwnProperty("dRtg") || val.hasOwnProperty("oRtg") ||
          val.hasOwnProperty("deltaOffPpp") || val.hasOwnProperty("deltaDefPpp")
        ),
        //(pick a field that all stat sets have - plus special case for: diagnostics)
      print: (val: any, serialize: ((val: any) => string), indent: ((val: any) => string)) => {
        const expectedFields = expectedFields1 || new Set(_.keys(exampleRawResponse));
        if (val.off_efg) {
          return "{\n" +
            _.chain(val).toPairs().filter(kv => {
              return !expectedFields.has(kv[0]) //derived field)
                || (!kv[1]?.hasOwnProperty("value") && exampleRawResponse[kv[0]]?.hasOwnProperty("value")) //missing value
                || (kv[1]?.hasOwnProperty("value") && _.keys(kv[1]).length > 1) //(has extra properties)
            }).sortBy(0).map(kv => {
              return indent(`${kv[0]}: ` + serialize(kv[1]))
            }).value().join("\n\n") +
          "}";
        } else if (val.hasOwnProperty("dRtg") || val.hasOwnProperty("oRtg")) {
          return JSON.stringify(_.pick(val, [
            "oRtg", "adjORtg", "adjORtgPlus", "dRtg", "adjDRtg", "adjDRtgPlus"
          ]), null , 3);
        } else if (val.hasOwnProperty("deltaOffPpp") || val.hasOwnProperty("deltaDefPpp")) {
          return JSON.stringify(_.pick(val, [
            "deltaOffPpp", "deltaOffEfg", "deltaOffAdjEff", "deltaDefPpp", "deltaDefEfg", "deltaDefAdjEff"
          ]), null , 3);
        } else {
          return "???";
        }
      }
    }
  }


  /** From easier to read templates builds a bucket list */
  static buildResponseFromTemplateLineup(input: Record<string, any>) {
    const lineupAsObj = SampleDataUtils.buildResponseFromTemplateTeam(input);
    return _.toPairs(lineupAsObj).map(kv => {
      return { key: kv[0], ...(kv[1] as Record<string, any>) };
    })
  }

  /** From easier to read templates builds a bucket list */
  static buildResponseFromTemplateTeam(input: Record<string, any>) {
    const teamAsObj = _.transform(_.toPairs(input), (acc, inElKv) => {
      _.forEach(_.toPairs(inElKv[1] as Record<string, any>), (kv) => {
        _.forEach(_.toPairs(kv[1] as Record<string, any>), (deepKv) => {
          if (inElKv[0] != "main") {
            _.set(acc, `${kv[0]}.${deepKv[0]}${inElKv[0]}.value`, deepKv[1]);
          } else {
            _.set(acc, `${deepKv[0]}.${kv[0]}`, deepKv[1]);
          }
        });
      });
    }, {} as Record<string, any>);
    return teamAsObj;
  }

  /** From easier to read templates builds a bucket list */
  static buildResponseFromTemplatePlayer(input: Record<string, any>) {
    const flatPlayersAsObj = SampleDataUtils.buildResponseFromTemplateTeam(input);
    return _.transform(["baseline", "on", "off"], (acc, v) => {
      acc[v] = { player: {
        buckets: _.chain(flatPlayersAsObj).toPairs().filter(kv => {
          return kv[0].indexOf(v) == 0;
        }).map(kv => {
          const key = kv[0].substring(v.length + 2); //(+2 for ": ")
          return { ...(kv[1]), key: key }
        }).value()
      } };
    }, {} as Record<string, any>);
  }

  /** Builds an easier to manipulate template for the different responses, from an actual response */
  static buildTemplateFromResponseTeam(input: Record<string, Record<string, any>>) {
    return SampleDataUtils.buildTemplateFromResponseLineup(
      _.chain(input).toPairs().map((kv: [string, any]) => {
        return { key: kv[0], ...(kv[1]) }
      }).value()
    );
  }

  /** Builds an easier to manipulate template for the different responses, from an actual response */
  static buildTemplateFromResponsePlayer(input: Record<string, Record<string, any>>) {
    return SampleDataUtils.buildTemplateFromResponseLineup(
      _.chain(input).toPairs().flatMap((kv: [string, any]) => {
        const players = (kv[1]?.player?.buckets || []) as Array<Record<string, any>>;
        return players.map(p => { return { ...p, key: `${kv[0]}: ${p.key}` } });
      }).value()
    );
  }

  /** Builds an easier to manipulate template for the different responses, from an actual response */
  static buildTemplateFromResponseLineup(input: Array<Record<string, any>>, filter?: Set<String>) {
    const preSorted = _.transform(input, (acc, inEl) => {
      const recordId = inEl?.key || "unknown";
      const injectInfo = (k: string, prefix: string) => {
        const topLevelKey = k.substring(prefix.length);
        const realPrefix = prefix || "main";
        if (prefix) {
          const existingObj = _.get(acc, `${topLevelKey}.${recordId}`) || {};
          existingObj[realPrefix] = inEl[k]?.value;
          const sortedObj = _(existingObj).toPairs().sortBy(0).reverse().fromPairs().value()
          _.set(acc, `${topLevelKey}.${recordId}`, sortedObj);
        } else {
          _.set(acc, `main.${topLevelKey}.${recordId}`, inEl[k]);
        }
      };

      if (!filter || filter.has(recordId)) {
        _.keys(inEl).filter(k => k != "key").forEach(k => {
          const stat = _.find([
            "oppo_total_off_", "oppo_total_def_",
            "oppo_off_", "oppo_def_",

            "team_total_off_", "team_total_def_",
            "team_off_", "team_def_",

            "total_off_", "total_def_",
            "off_", "def_"
          ], prefix => k.indexOf(prefix) == 0);
          if (stat) {
            injectInfo(k, stat);
          } else {
            injectInfo(k, "");
          }
        });
      }
    }, {} as Record<string, any>);

    // Sort the keys into some sort of logical order
    return _.chain(preSorted).toPairs().sortBy([
      (kv: [ string , any ]) => {
        const key = kv[0];
        if (key == "main") {
          return 100;
        } else if ((key.indexOf("assist") >= 0) || (key.indexOf("to") >= 0) || (key.indexOf("ast") >= 0)) {
          return 4;
        } else if ((key.indexOf("ft") == 0) || (key.indexOf("fg") == 0) || (key == "efg")) {
          return 1;
        } else if ((key.indexOf("2p") == 0) || (key.indexOf("3p") == 0)) {
          return 2;
        } else if ((key.indexOf("rb") > 0)) {
          return 3;
        } else if ((key.indexOf("adj") >= 0)) {
          return 5;
        } else { //misc other stuff
          return 6;
        }
      },
      (kv: [ string , any ]) => {
        return kv[0];
      },
    ]).fromPairs().value();
  }
}
