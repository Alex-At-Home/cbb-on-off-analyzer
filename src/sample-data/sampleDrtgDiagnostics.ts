export const sampleDrtgDiagnostics = {
  "stl": 30,
  "blk": 5,
  "drb": 94,
  "pfPct": 0.2,
  "playerRtg": 142.82324406037372,
  "playerDelta": 8.134297736197055,
  "scPossConceded": 0.6341530662717683,
  "noShotCredit": 32.188268933121634,
  "reboundCredit": 36.07039815349106,
  "missFtCredit": 6.070287539936104,
  "stopsIndPct": 0.21044437889736348,
  "stopsTeamPct": 0.15540255483086815,
  "teamBlk": 60,
  "oppoPts": 1804,
  "oppoPoss": 1766,
  "oppoFga": 1429,
  "oppoFgm": 534,
  "oppoFtm": 426,
  "oppoFta": 626,
  "oppoFtPoss": 297.34999999999997,
  "oppoTov": 59,
  "teamStl": 150,
  "opponentOrbPct": 0.2708757637474542,
  "opponentFgPct": 0.3736878936319104,
  "teamOrbCreditToDefender": 0.2724649000411893,
  "teamOrbCreditToRebounder": 0.16965277015673305,
  "teamDvsRebCredit": 0.616272360069244,
  "oppoFgMiss": 835,
  "oppoNonStlTov": -91,
  "teamMissWeight": 0.4376537866243272,
  "oppoFtPct": 0.6805111821086262,
  "oppoFtHitOnePlus": 266.99856230031946,
  "oppoProbFtHitOnePlus": 0.8979268952423726,
  "oppoScPoss": 800.9985623003195,
  "oppoPtsPerScore": 2.2521888114496065,
  "teamRtg": 102.15175537938845,
  "adjDRtg": 102.04282992219363,
  "adjDRtgPlus": 0.4085659844387266,
  "avgEff": 100,
  "dRtg": 110.2860531155855,
  "offSos": 108.0781993204983
 };
