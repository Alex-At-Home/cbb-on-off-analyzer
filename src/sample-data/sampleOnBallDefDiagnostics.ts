/** Manually built by copying the snapshot from src/utils/stats/__tests__/RatingUtils.test.ts */
export const sampleOnBallDefDiagnostics =
{
   "playerRebCredit": 0.09664054373332856,
   "teamRebCredit": 0.14616882239665946,
   "comboRebCredit": 0.2948067270155614,
   "playerTargetPoss": 185.586,
   "targetedPct": 0.12134831460674157,
   "playerPtsPerScore": 2.4291266575217194,
   "onBallStopCredit": 0.4668039520787198,
   "offBallPoss": 1572.7865168539327,
   "offBallPts": 1539.0286760055637,
   "otherRebounds": 631.6805699093197,
   "otherRebCredit": 0.14474344317512738,
   "offBallStopCredit": 0.39516433361719344,
   "onVsOffBallWeight": 0.4,
   "weightedPtsPerScore": 2.126826397164934,
   "onBallCreditWeight": 0.04853932584269663,
   "offBallCreditWeight": 0.13179775280898876,
   "comboBallStopCredit": 0.8290444280732188,
   "unadjDRtg": 103.9658869175629,
   "weightedClassicDRtgMean": 98.84955759282161,
   "weightedUnadjDRtgMean": 107.19011038664136,
   "uncategorizedAdjustment": -7.10724422350558,
   "adjustedPossPct": 0.08000000000000002,
   "dRtg": 96.85864269405732,
   "adjDRtg": 89.6509562344713,
   "adjDRtgPlus": -2.069808753105741
};
