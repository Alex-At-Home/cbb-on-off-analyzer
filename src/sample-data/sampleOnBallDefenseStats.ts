export const sampleOnBallDefenseStats =
[
  {
     "code": "totals",
     "title": "Team Name",
     "pts": 1554,
     "plays": 1780,
     "scorePct": 39.1,
     "tovPct": 16.5,
     "fgMiss": 769,
     "totalPts": -1,
     "totalScorePct": -1,
     "totalPlays": -1,
     "uncatPts": -1,
     "uncatPlays": -1,
     "uncatScorePct": -1,
     "uncatPtsPerScPlay": -1
  },  [
   {
      "code": "Code1",
      "title": "#1 Player1",
      "pts": 170,
      "plays": 216,
      "scorePct": 32.4,
      "tovPct": 15.7,
      "fgMiss": 111,
      "totalPts": -1,
      "totalScorePct": -1,
      "totalPlays": -1,
      "uncatPts": -1,
      "uncatPlays": -1,
      "uncatScorePct": -1,
      "uncatPtsPerScPlay": -1
   },
   {
      "code": "Code2",
      "title": "#02 Player2",
      "pts": 159,
      "plays": 170,
      "scorePct": 41.2,
      "tovPct": 11.8,
      "fgMiss": 79,
      "totalPts": -1,
      "totalScorePct": -1,
      "totalPlays": -1,
      "uncatPts": -1,
      "uncatPlays": -1,
      "uncatScorePct": -1,
      "uncatPtsPerScPlay": -1
   },
   {
      "code": "Code3",
      "title": "#13 Player3",
      "pts": 151,
      "plays": 169,
      "scorePct": 37.3,
      "tovPct": 16,
      "fgMiss": 78,
      "totalPts": -1,
      "totalScorePct": -1,
      "totalPlays": -1,
      "uncatPts": -1,
      "uncatPlays": -1,
      "uncatScorePct": -1,
      "uncatPtsPerScPlay": -1
   },
   {
      "code": "Code4",
      "title": "#99 Player4",
      "pts": 109,
      "plays": 107,
      "scorePct": 39.3,
      "tovPct": 12.1,
      "fgMiss": 52,
      "totalPts": -1,
      "totalScorePct": -1,
      "totalPlays": -1,
      "uncatPts": -1,
      "uncatPlays": -1,
      "uncatScorePct": -1,
      "uncatPtsPerScPlay": -1
   },
   {
      "code": "Code5",
      "title": "#21 Player5",
      "pts": 103,
      "plays": 148,
      "scorePct": 31.8,
      "tovPct": 14.9,
      "fgMiss": 79,
      "totalPts": -1,
      "totalScorePct": -1,
      "totalPlays": -1,
      "uncatPts": -1,
      "uncatPlays": -1,
      "uncatScorePct": -1,
      "uncatPtsPerScPlay": -1
   },
   {
      "code": "Code6",
      "title": "#66 Player6",
      "pts": 52,
      "plays": 88,
      "scorePct": 26.1,
      "tovPct": 11.4,
      "fgMiss": 54,
      "totalPts": -1,
      "totalScorePct": -1,
      "totalPlays": -1,
      "uncatPts": -1,
      "uncatPlays": -1,
      "uncatScorePct": -1,
      "uncatPtsPerScPlay": -1
   },
   {
      "code": "Code7",
      "title": "Player 7",
      "pts": 76,
      "plays": 95,
      "scorePct": 32.6,
      "tovPct": 14.7,
      "fgMiss": 50,
      "totalPts": -1,
      "totalScorePct": -1,
      "totalPlays": -1,
      "uncatPts": -1,
      "uncatPlays": -1,
      "uncatScorePct": -1,
      "uncatPtsPerScPlay": -1
   } ]
];
