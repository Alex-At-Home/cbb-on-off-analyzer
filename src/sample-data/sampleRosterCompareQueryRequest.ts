export const sampleRosterCompareQueryRequest =                 {
         "_source": {
            "includes": [],
            "excludes": []
         },
         "size": 0,
         "aggregations": {
            "tri_filter": {
               "filters": {
                  "filters": {
                     "off": {
                        "query_string": {
                           "query": "(players.id:(NOT *)) AND (players.id:(base))"
                        }
                     },
                     "on": {
                        "query_string": {
                           "query": "(players.id:(NOT *)) AND (players.id:(base))"
                        }
                     },
                     "baseline": {
                        "query_string": {
                           "query": "players.id:(base)"
                        }
                     }
                  }
               },
               "aggregations": {
                  "global_poss_count": {
                     "sum": {
                        "field": "team_stats.num_possessions"
                     }
                  },
                  "player": {
                     "terms": {
                        "field": "players.id.keyword",
                        "size": 100
                     },
                     "aggregations": {
                        "poss_count": {
                           "sum": {
                              "field": "team_stats.num_possessions"
                           }
                        },
                        "sort": {
                           "bucket_sort": {
                              "from": 0,
                              "sort": [
                                 {
                                    "poss_count": {
                                       "order": "desc"
                                    }
                                 }
                              ],
                              "size": 100
                           }
                        }
                     }
                  }
               }
            }
         },
         "query": {
            "bool": {
               "filter": [],
               "must_not": [],
               "should": [],
               "minimum_should_match": 0,
               "must": [
                  {
                     "script": {
                        "script": {
                           "source": "if (params.kp.isEmpty()) return true;\ndef kp_name = params.pbp_to_kp[doc[\"opponent.team.keyword\"].value];\nif (kp_name == null) {\n   kp_name = doc[\"opponent.team.keyword\"].value;\n} else {\n   kp_name = kp_name.pbp_kp_team\n}\ndef oppo = params.kp[kp_name];\nif (oppo != null) {\n def kp_rank = oppo['stats.adj_margin.rank'];\n def game_filter = params.game_filter.game_filter;\n //TODO: high major\n return (kp_rank >= game_filter.min_kp) && (kp_rank <= game_filter.max_kp);\n} else {\n return false;\n}\n\n",
                           "lang": "painless",
                           "params": {
                              "pbp_to_kp": {
                                 "name1": "name1b"
                              },
                              "kp": {
                                 "team": {
                                    "stats": 0
                                 }
                              },
                              "game_filter": {
                                 "game_filter": {
                                    "min_kp": 10,
                                    "max_kp": 100
                                 }
                              }
                           }
                        }
                     }
                  },
                  {
                     "term": {
                        "team.team.keyword": "undefined"
                     }
                  }
               ]
            }
         }
      };
