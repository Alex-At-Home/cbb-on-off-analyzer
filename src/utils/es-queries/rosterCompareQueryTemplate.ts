import { commonTeamQuery } from "./commonTeamQuery";
import { commonOnOffBaseQuery } from "./commonOnOffBaseQuery";
import { GameFilterParams } from "../FilterModels";

export const rosterCompareQuery = function(params: GameFilterParams, publicEfficiency: any, lookup: any) {
  return {
     "_source": {
        "includes": [],
        "excludes": []
     },
     "size": 0,
     "aggregations": {
        "tri_filter": {
          "filters": commonOnOffBaseQuery(params),
           "aggregations": {
              "global_poss_count": {
                 "sum": {
                    "field": "team_stats.num_possessions"
                 }
              },
              "player": {
                 "terms": {
                    "field": "players.id.keyword",
                    "size": 100
                 },
                 "aggregations": {
                    "poss_count": {
                       "sum": {
                          "field": "team_stats.num_possessions"
                       }
                    },
                    "sort": {
                       "bucket_sort": {
                          "from": 0,
                          "sort": [
                             {
                                "poss_count": {
                                   "order": "desc"
                                }
                             }
                          ],
                          "size": 100
                       }
                    }
                 }
              }
           }
        }
     },
     "query": commonTeamQuery(params, publicEfficiency, lookup)
  };
}
