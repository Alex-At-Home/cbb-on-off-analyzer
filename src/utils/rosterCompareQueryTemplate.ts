import { ncaaToKenpomLookup } from "./ncaaToKenpomLookup"
import { publicKenpomEfficiency2018 } from "./publicKenpomEfficiency2018"

export const rosterCompareQuery2018 = function(params: any) {
  return { //TODO: handle multiple years
     "_source": {
        "includes": [],
        "excludes": []
     },
     "size": 0,
     "aggregations": {
        "tri_filter": {
          "filters": {
             "filters": {
                "off": {
                   "query_string": {
                      "query": `players.id:((${params.offQuery || "*"}) AND (${params.baseQuery || "*"}))`
                   }
                },
                "on": {
                   "query_string": {
                     "query": `players.id:((${params.onQuery || "*"}) AND (${params.baseQuery || "*"}))`
                   }
                },
                "baseline": {
                   "query_string": {
                      "query": `players.id:(${params.baseQuery || "*"})`
                   }
                }
             }
          },
           "aggregations": {
              "global_poss_count": {
                 "sum": {
                    "field": "team_stats.num_possessions"
                 }
              },
              "player": {
                 "terms": {
                    "field": "players.id.keyword",
                    "size": 100
                 },
                 "aggregations": {
                    "poss_count": {
                       "sum": {
                          "field": "team_stats.num_possessions"
                       }
                    },
                    "sort": {
                       "bucket_sort": {
                          "from": 0,
                          "sort": [
                             {
                                "poss_count": {
                                   "order": "desc"
                                }
                             }
                          ],
                          "size": 100
                       }
                    }
                 }
              }
           }
        }
     },
     "query": {
        "bool": {
           "filter": [],
           "must_not": [],
           "should": [],
           "must": [
              {
                 "script": {
                    "script": {
                       "source": "def kp_name = params.pbp_to_kp[doc[\"opponent.team.keyword\"].value];\nif (kp_name != null) {\ndef oppo = params.kp[kp_name.pbp_kp_team];\nif (oppo != null) {\n   def kp_rank = oppo['stats.adj_margin.rank'];\n   def game_filter = params.game_filter.game_filter;\n   //TODO: high major\n   return (kp_rank >= game_filter.min_kp) && (kp_rank <= game_filter.max_kp);\n} else {\n   return false;\n}\n} else {\n   return false;\n}\n",
                       "lang": "painless",
                       "params": {
                          "pbp_to_kp": ncaaToKenpomLookup,
                          "kp": publicKenpomEfficiency2018,
                          "game_filter": {
                             "game_filter": {
                               "min_kp": Number(params.minRank),
                               "max_kp": Number(params.maxRank)
                             }
                          }
                       }
                    }
                 }
              },
              {
                 "query_string": {
                    "query": "*"
                 }
              }
           ]
        }
     }
  };
}
