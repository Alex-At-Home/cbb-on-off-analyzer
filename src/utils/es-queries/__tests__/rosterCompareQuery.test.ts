
import { rosterCompareQuery } from "../rosterCompareQueryTemplate";
import { sampleRosterCompareQueryRequest } from "../../../sample-data/sampleRosterCompareQueryRequest";

describe("rosterCompareQuery", () => {
  test("rosterCompareQuery  ", () => {

    const test = rosterCompareQuery({
      team: "TestTeam", year: "2019",
      minRank: "10", maxRank: "100",
      baseQuery: "base",
    }, 0, { "team": { "stats": 0 } }, {
      "name1": "name1b"
    });
    expect(test).toEqual(sampleRosterCompareQueryRequest);
  });
});
