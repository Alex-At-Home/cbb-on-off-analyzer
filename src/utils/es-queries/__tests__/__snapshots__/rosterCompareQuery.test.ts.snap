// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rosterCompareQuery rosterCompareQuery   1`] = `
{
   "runtime_mappings": {
      "vs_high_major": {
         "type": "boolean",
         "script": {
            "source": "if (0 != doc['common_lookup'].size()) emit((doc['common_lookup'].value & 1) > 0)"
         }
      },
      "in_conf": {
         "type": "boolean",
         "script": {
            "source": "if (0 != doc['common_lookup'].size()) emit((doc['common_lookup'].value & 2) > 0)"
         }
      },
      "vs_rank": {
         "type": "long",
         "script": {
            "source": "if (0 != doc['common_lookup'].size()) emit((doc['common_lookup'].value >> 2) & 511)"
         }
      },
      "vs_3p": {
         "type": "double",
         "script": {
            "source": "if (0 != doc['common_lookup'].size()) {\\n            def _3p = (doc['common_lookup'].value >> 11)  & 1023;\\n            if (_3p > 0) emit(0.1*_3p);\\n          }"
         }
      },
      "vs_adj_off": {
         "type": "double",
         "script": {
            "source": "if (0 != doc['common_lookup'].size()) emit(0.1*((doc['common_lookup'].value >> 21) & 2047))"
         }
      },
      "vs_adj_def": {
         "type": "double",
         "script": {
            "source": "if (0 != doc['common_lookup'].size()) emit(0.1*((doc['common_lookup'].value >> 32) & 2047))"
         }
      },
      "common_lookup": {
         "type": "long",
         "script": {
            "source": "\\n            if (!params.kp_info.isEmpty()) {\\n              def raw_name = doc[params.field_name].value;\\n              def kp_name = params.pbp_to_kp[raw_name];\\n              if (kp_name == null) {\\n                 kp_name = raw_name;\\n              } else {\\n                 kp_name = kp_name.pbp_kp_team;\\n              }\\n              def oppo = params.kp_info[kp_name];\\n              if (oppo != null) {\\n                def vs_high_major = oppo[\\"is_high_major\\"] ?: 1;\\n                if (vs_high_major instanceof String) {\\n                  vs_high_major = 0;\\n                }\\n                def oppo_conf = oppo[\\"conf\\"];\\n                def team_conf = params.conf;\\n                if (team_conf == \\"*\\") {      \\n                  def raw_team_name = doc[\\"team.team.keyword\\"].value; \\n                  def team_kp_name = params.pbp_to_kp[raw_team_name];\\n                  if (team_kp_name == null) {\\n                    team_kp_name = raw_team_name;\\n                  } else {\\n                    team_kp_name = team_kp_name.pbp_kp_team;\\n                  }\\n                  def team_kp = params.kp_info[team_kp_name];\\n                  if (team_kp != null) {\\n                    team_conf = team_kp[\\"conf\\"];\\n                  }\\n                }\\n                def in_conf = team_conf.equals(oppo_conf);\\n                def hca = 0;\\n                if (doc[\\"location_type.keyword\\"].size() > 0) {\\n                  if (doc[\\"location_type.keyword\\"].value == \\"Neutral\\") {\\n                    hca = 25;\\n                  } else if (doc[\\"location_type.keyword\\"].value == \\"Away\\") {\\n                    hca = 50;\\n                  }\\n                }\\n                def margin_rank = Math.max(1, (oppo[\\"stats.adj_margin.rank\\"] ?: 0) - hca);\\n\\n                def _3p = oppo[\\"stats.off._3p_pct.value\\"] ?: 0.0;\\n                _3p = (_3p * 10).longValue() & 1023;\\n\\n                def adj_off = oppo[\\"stats.adj_off.value\\"] ?: 0;\\n                adj_off = (adj_off * 10).longValue() & 2047;\\n\\n                def adj_def = oppo[\\"stats.adj_def.value\\"] ?: 0;\\n                adj_def = (adj_def * 10).longValue() & 2047;\\n\\n                long returnVal = 0L | (vs_high_major << 0)\\n                  | ((in_conf ? 1 : 0) << 1)\\n                  | ((margin_rank.longValue() & 511) << 2)\\n                  | (_3p.longValue() << 11)\\n                  | (adj_off.longValue() << 21)\\n                  | (adj_def.longValue() << 32)\\n                  ;\\n                emit(returnVal);\\n              }\\n            }\\n          ",
            "lang": "painless",
            "params": {
               "pbp_to_kp": {
                  "name1": "name1b"
               },
               "field_name": "opponent.team.keyword",
               "kp_info": {
                  "team": {
                     "stats": 0
                  }
               },
               "conf": ""
            }
         }
      }
   },
   "_source": {
      "includes": [],
      "excludes": []
   },
   "size": 0,
   "aggregations": {
      "tri_filter": {
         "filters": {
            "filters": {
               "off": {
                  "query_string": {
                     "query": "(players.id:(NOT *)) AND (players.id:(base))"
                  }
               },
               "on": {
                  "query_string": {
                     "query": "(players.id:(NOT *)) AND (players.id:(base))"
                  }
               },
               "baseline": {
                  "query_string": {
                     "query": "players.id:(base)"
                  }
               }
            }
         },
         "aggregations": {
            "global_poss_count": {
               "sum": {
                  "field": "team_stats.num_possessions"
               }
            },
            "player": {
               "terms": {
                  "field": "players.id.keyword",
                  "size": 100
               },
               "aggregations": {
                  "poss_count": {
                     "sum": {
                        "field": "team_stats.num_possessions"
                     }
                  },
                  "sort": {
                     "bucket_sort": {
                        "from": 0,
                        "sort": [
                           {
                              "poss_count": {
                                 "order": "desc"
                              }
                           }
                        ],
                        "size": 100
                     }
                  }
               }
            }
         }
      }
   },
   "query": {
      "bool": {
         "filter": [],
         "must_not": [],
         "should": [],
         "minimum_should_match": 0,
         "must": [
            {
               "term": {
                  "team.team.keyword": "TestTeam"
               }
            },
            {
               "query_string": {
                  "query": "vs_rank:[10 TO 100]"
               }
            }
         ]
      }
   }
}
`;
